internal/repository/index.go

```diff
@@ -19,40 +19,33 @@ import (
 // Hence the index data structure defined here is one of the main contributions
 // to the total memory requirements of restic.
 //
-// We use two maps to store each index entry.
-// The first map stores the first entry of a blobtype/blobID
-// The key of the map is a BlobHandle
-// The entries are the actual index entries.
-// In the second map we store duplicate index entries, i.e. entries with same
-// blobtype/blobID
+// We store the index entries in indexMaps. In these maps, entries take 56
+// bytes each, plus 8/4 = 2 bytes of unused pointers on average, not counting
+// malloc and header struct overhead and ignoring duplicates (those are only
+// present in edge cases and are also removed by prune runs).
+//
 // In the index entries, we need to reference the packID. As one pack may
 // contain many blobs the packIDs are saved in a separate array and only the index
 // within this array is saved in the indexEntry
 //
-// To compute the needed amount of memory, we need some assumptions.
-// Maps need an overhead of allocated but not needed elements.
-// For computations, we assume an overhead of 50% and use OF=1.5 (overhead factor)
-// As duplicates are only present in edge cases and are also removed by prune runs,
-// we assume that there are no significant duplicates and omit them in the calculations.
-// Moreover we asssume on average a minimum of 8 blobs per pack; BP=8
+// We assume on average a minimum of 8 blobs per pack; BP=8.
 // (Note that for large files there should be 3 blobs per pack as the average chunk
 // size is 1.5 MB and the minimum pack size is 4 MB)
 //
 // We have the following sizes:
-// key: 32 + 1 = 33 bytes
-// indexEntry:  8 + 4 + 4 = 16 bytes
+// indexEntry:  56 bytes  (on amd64)
 // each packID: 32 bytes
 //
 // To save N index entries, we therefore need:
-// N * OF * (33 + 16) bytes + N * 32 bytes / BP = N * 78 bytes
+// N * (56 + 2) bytes + N * 32 bytes / BP = N * 62 bytes.
```

Comment:
@aawsome, does this strike you as correct? I feel I might have made a mistake in estimating the average load factor.

Reply:
I think that the calculations of the "old" implementation cannot be taken 1:1 to see the memory usage of this "new" implementation.

For the old implementation I argued that index entries of 49 bytes needed to be stored in a golang standard hashmap. This is usually partly filled and I assumed 2/3 filling (or overhead factor OF=1.5). What is completely neglected is the extra overhead that comes from the map, i.e. the pointers to each bucket and maybe even more overhead when the map is growing. So the `N * OF * 49 bytes` are the contents within the map but without the organizational overhead the map uses itself.

For this "new" implementaion, speaking in the same context would be:
We store 48 bytes per entry and there is no overhead. :smile: This would lead to `N * 52 bytes` total usage.

However, we also do have some "organizational" overhead here: 1 pointer (8 bytes) per entry plus 1 pointer per bucket. The number of buckets is between `N/2` and `N/4`, so in total the overhead is smaller than `N * 12 bytes` (+ some overhead not depending on the number of entries, but this we can neglect).

So for this new implementation the total memory is less than `64 bytes` per entry *including any overhead*.